맛집 데이터 관리, 조회

kingseungil opened the pull request:
#️⃣ 연관된 이슈
#4, #5

📝 작업 내용
- openFeign을 이용하여 외부 데이터 가져오기
- 주기적으로(한달마다 스케쥴링) mysql에 데이터 저장(수정된 데이터가 있으면 update, 새로운 데이터면 save)
- mysql - es 매핑 (1시간마다 스케쥴링, mysql 데이터의 수정일시가 한시간 이내인것만 인덱싱)
- es를 활용하여 맛집 정보들 조회 (전체 조회, 지역별 조회, 내 위치 주변 조회)
- 내 위치 주변 조회는 우선 회원가입 시 등록한 region을 기준으로 조회 중입니다. (위도, 경도 데이터가 있어서 추후에 변경 가능)
- 각 조회 시 rating, reviewCount로 정렬 가능
- mysql id 컬럼으로 맛집 상세조회

💬 리뷰 요구사항

kingseungil committed:
chore: openfeign 종속성 추가 (#4)

kingseungil committed:
feat: openfeign 설정 (#4)

- 외부 API 가져오는 인터페이스 추가
- dto 설정
- decode config 설정 (응답값 추출 - data 부분)

kingseungil committed:
feat: docker-compose 추가 (#4)

- mysql, redis, elasticsearch
- 추후에 백엔드 서버도 포함해야함

kingseungil committed:
feat: elasticsearch 초기설정 (#4)

kingseungil committed:
feat: openAPI 데이터 DB에 저장 (#4)

- 한달에 한번씩 스케쥴링

kingseungil committed:
feat: es docker nori 플러그인 추가 (#4)

- 한글 검색 위해서 플러그인 추가

kingseungil committed:
feat: es-mysql 매핑 (#4)

- 한시간 단위로 mysql 데이터에서 업데이트된 데이터만 es에 upsert

kingseungil committed:
fix: openAPI 데이터 - mysql 저장 수정 (#4)

- 기존에 스케쥴링으로 데이터를 받아와서 디비에 저장해주고 있었는데, 동일한 데이터가 들어오는 상황 발생
- 디비에 저장할 때 이미 존재하는 데이터인지 확인하는 로직 추가
- Restaurant 테이블에 name,address 복합 인덱스 설정 (동일한 이름의 매장이 존재하므로 주소와 함께 판단하기 위함)

kingseungil committed:
feat: 지번 주소 필드 추가 (#4)

kingseungil committed:
feat: 식당 조회 api (#4)

- 전체 식당 조회
- 주소로 식당 조회

kingseungil committed:
feat: 내 주변 맛집 조회 (#5)

kingseungil committed:
doc: READMD.md

- erd 수정(컬럼 추가 - 지번주소)

kingseungil committed:
feat: 매장 상세 조회 api (#5)

kingseungil committed:
test: 매장 조회 api (#5)

kingseungil committed:
feat: 매장데이터에 rating, review count 추가 (#5)

kingseungil committed:
feat: 각 조회 시 별점 순, 리뷰 개수 순으로 정렬기능 추가 (#5)

kingseungil committed:
test: 각 조회 시 별점 순, 리뷰 개수 순으로 정렬기능 추가 (#5)

kingseungil committed:
test: 각 조회 시 별점 순, 리뷰 개수 순으로 정렬기능 추가 (#5)

mingming-mentor commented on src/main/java/matgo/restaurant/application/RestaurantService.java:
```diff
@@ -0,0 +1,149 @@
+        List<String> restaurantAddresses = restaurants.stream()
+                                                      .map(Restaurant::getAddress)
+                                                      .toList();
+        List<Restaurant> existingRestaurants = restaurantRepository.findByNameInAndAddressIn(restaurantNames,
+          restaurantAddresses);
```
공공데이터에 식당 ID(PK) 값이 있어보이는데요, 이 값을 이용하여 식당 구분을 하지 않고 이름과 주소로 구분하는 이유가 있을까요? 폐업 등을 고려하여 이렇게 설계하신걸까요?

kingseungil replied:
@mingming-mentor 님
저는 공공데이터에서 제공해주는 ID값을 사용하지않아서 유일성을 보장해주는 이름과 주소로 체크를 했습니다.
처음엔 이름으로만 했는데 중복된 이름의 매장들이 있어서 주소까지 함께 체크해줬습니다.

위 방식보다 공공데이터에서 주는 ID값을 제 DB의 별도의 필드에 저장해서 그 값으로 체크를 해주는게 더 좋을까요?

mingming-mentor replied:
@kingseungil 님 네 아무래도 ID 값이 PK이니 그 값을 활용하는 편이 좀더 좋을 것 같긴합니다 .

mingming-mentor commented on src/main/java/matgo/restaurant/application/RestaurantService.java:
```diff
@@ -0,0 +1,149 @@
+    // 한달에 한번씩 실행
+    @Scheduled(cron = "0 0 0 1 * *")
+    public void fetchAndSaveRestaurants() {
+        List<RestaurantData> restaurantDataList = jeonjuRestaurantClient.getRestaurants(1, MAX_COUNT, key);
```
최대 4000개만 가지고 오는 것일까요? 그 이상에 대해서는 대응은 별도로 되지 않나요?

kingseungil replied:
@mingming-mentor 님
openAPI 데이터가 4000개가 좀 안돼서 저렇게 잡았었습니다.
```java
public void fetchAndSaveRestaurants() {
        int page = 1;
        while (true) {
            log.info("page : {}", page);
            List<RestaurantData> restaurantDataList = fetchRestaurantsFromAPI(page);
            if (restaurantDataList.isEmpty()) {
                break;
            }
            List<Restaurant> restaurants = convertToRestaurants(restaurantDataList);
            Map<String, Restaurant> existingRestaurantsMap = getExistingRestaurantsMap(restaurants);
            List<Restaurant> newRestaurants = getNewRestaurants(restaurants, existingRestaurantsMap);

            saveRestaurants(newRestaurants);
            log.info("fetch and save restaurants success");
            indexingToES();

            page++;
        }
    }
```
이런식으로 MAX_COUNT에 의존하지 않고 있는 데이터 다 가져올수 있게 수정해봤습니다.

mingming-mentor replied:
@kingseungil 님, API가 count (=전체 데이터의 수)를 제공한다면 이를 이용하면 어떨까 싶어요. while 문의 경우 무한루프에 빠질수 있기 때문에 되도록 사용을 지양하는 편이 좋긴합니다.

mingming-mentor commented on src/main/java/matgo/restaurant/application/RestaurantService.java:
```diff
@@ -0,0 +1,149 @@
+            Optional<Restaurant> optionalRestaurant
+              = existingRestaurants.stream()
+                                   .filter(
+                                     r -> r.getName().equals(restaurant.getName()) &&
+                                       r.getAddress().equals(restaurant.getAddress()))
```
이렇게 매번 filtering 하는 것보다 existingRestaurants를 name+address 한 값을 key로 하여 미리 map으로 만든뒤 restaurant의 name+address 값을 가지고 map에서 데이터를 체크해서 저장된 데이터가 있는지 없는지를 확인해보면 좋을 것 같아요

kingseungil replied:
@mingming-mentor 님
```java
private Map<String, Restaurant> getExistingRestaurantsMap(List<Restaurant> restaurants) {
        List<String> restaurantNames = restaurants.stream()
                                                  .map(Restaurant::getName)
                                                  .toList();
        List<String> restaurantAddresses = restaurants.stream()
                                                      .map(Restaurant::getAddress)
                                                      .toList();
        List<Restaurant> existingRestaurants = restaurantRepository.findByNameInAndAddressIn(restaurantNames,
          restaurantAddresses);

        Map<String, Restaurant> map = new HashMap<>();
        for (Restaurant existingRestaurant : existingRestaurants) {
            map.put(existingRestaurant.getName() + existingRestaurant.getAddress(), existingRestaurant);
        }
        return map;
    }
    
private List<Restaurant> getNewRestaurants(List<Restaurant> restaurants,
      Map<String, Restaurant> existingRestaurantsMap) {
        List<Restaurant> newRestaurants = new ArrayList<>();
        for (Restaurant restaurant : restaurants) {
            String key = restaurant.getName() + restaurant.getAddress();
            Restaurant existingRestaurant = existingRestaurantsMap.get(key);
            if (existingRestaurant != null) {
                existingRestaurant.update(restaurant);
            } else {
                newRestaurants.add(restaurant);
            }
        }
        return newRestaurants;
    }
```
좋은 피드백 감사합니다. 피드백 반영해서 수정해봤습니다

mingming-mentor replied:
@kingseungil 님 조금더 피드백 드려보자면 이렇게 Map을 만든뒤 put을 하는 코드보단 stream을 사용하여 map을 선언과 함께 할당하는 방식이 좀더 지향되고 있습니다. 참고부탁드려요

~~~java
// 기존
        Map<String, Restaurant> map = new HashMap<>();
        for (Restaurant existingRestaurant : existingRestaurants) {
            map.put(existingRestaurant.getName() + existingRestaurant.getAddress(), existingRestaurant);
        }

// 수정
existingRestaurants.stream().collect(Collectors.toMap(restaurant -> restaurant.getName() + 
restaurant.getAddress(), restaurant -> restaurant);
~~~

mingming-mentor commented on src/main/java/matgo/restaurant/application/RestaurantService.java:
```diff
@@ -0,0 +1,149 @@
+        // 1시간 이내에 수정된 식당만 elasticsearch에 저장
+        LocalDateTime oneHourAgo = LocalDateTime.now().minusHours(1);
+        List<Restaurant> restaurants = restaurantRepository.findByModifiedAtAfter(oneHourAgo);
+        if (restaurants.isEmpty()) {
+            throw new RestaurantException(UPDATABLE_RESTAURANT_NOT_FOUND);
+        }
```
이 부분 exception이 던져져도 괜찮나요? 업데이트 한 내역이 없다면 isEmpty = true 가 당연할것 같아서요, 에러 상황으로 판단해도 좋을지 궁금합니다

kingseungil replied:
@mingming-mentor 님
```java
if (restaurants.isEmpty()) {
            log.info("No restaurants updated in the last hour");
        } else {
            List<RestaurantSearch> restaurantSearches = restaurants.stream()
                                                                   .map(RestaurantSearch::from)
                                                                   .toList();
            restaurantSearchRepositoryImpl.bulkInsertOrUpdate(restaurantSearches);
            log.info("elasticsearch indexing success");
        }
```
에러가 아닌 로그만 찍는걸로 수정해봤습니당

mingming-mentor commented on src/main/java/matgo/restaurant/application/RestaurantService.java:
```diff
@@ -0,0 +1,149 @@
+    @Transactional
+    // 한시간에 한번씩 실행
+    @Scheduled(cron = "0 0 * * * *")
+    public void indexingToES() {
```
ES에 저장하는 부분이 유저가 맛집 등록요청시에 async로 동작하거나, 혹은 위에서 fetchAndSaveRestaurants를 할때 동작하는 것이 아닌 별도로 scheduling 된 이유가 있을까요? 변경내역이 바로바로 반영이 안될듯한데 문제 없을지 궁금합니다.

kingseungil replied:
@mingming-mentor 님
맛집 등록되거나 리뷰작성했을 때 실시간 동기화를 진행할텐데, 혹시 빠진 부분을 위해 더블체크용으로 만들었습니다. `fetchAndSaveRestaurants` 메서드는 db에 저장하는 용도라서 이 부분에서 인덱싱을 해주진 않았었습니다.
하지만 `fetchAndSaveRestaurants`에서 디비 저장 후 인덱싱까지 해주면 더 좋을 것 같아서 수정해봤습니다

mingming-mentor reviewed (approved):
수고 많으셨습니다

kingseungil committed:
refactor: 피드백 반영 (#5)

- fetchAndSaveRestaurants 메서드 리팩토링
- indexingToES 메서드 리팩토링

mingming-mentor commented:
수고 많으셨어요!

kingseungil committed:
refactor: 피드백 반영 (#5)

- fetchAndSaveRestaurants 메서드 리팩토링
- 외부API에서 제공해주는 ID값을 이용하여 중복체크 (external_id  필드 추가)

kingseungil committed:
test: 피드백 반영 (#5)

- fetchAndSaveRestaurants 메서드 리팩토링
- 외부API에서 제공해주는 ID값을 이용하여 중복체크 (external_id  필드 추가)
